src/charm.py

```diff
@@ -49,68 +48,96 @@ def __init__(self, *args):
+            "remote-store-insecure": "false",
+        }
 
-    def _on_store_endpoints_changed(self, event):
-        """Generate store config."""
-        self._configure_store(event)
+    # === EVENT HANDLERS === #
 
-    def _install(self, _):
+    def _on_install(self, _):
         """Install dependencies for Parca Agent and ensure initial configs are written."""
         self.unit.status = ops.MaintenanceStatus("installing parca-agent")
         try:
             self.parca_agent.install()
-            self.unit.set_workload_version(self.parca_agent.version)
         except snap.SnapError as e:
-            self.unit.status = ops.BlockedStatus(str(e))
+            logger.exception("Failed to install parca-agent snap %s", str(e))
 
-    def _upgrade_charm(self, _):
+    def _on_upgrade_charm(self, _):
         """Ensure the snap is refreshed (in channel) if there are new revisions."""
         self.unit.status = ops.MaintenanceStatus("refreshing parca-agent")
         try:
             self.parca_agent.refresh()
         except snap.SnapError as e:
-            self.unit.status = ops.BlockedStatus(str(e))
-
-    def _update_status(self, _):
-        """Handle the update status hook (on an interval dictated by model config)."""
-        # Ensure the hold is extended to make sure the snap never auto-refreshes
-        # out of our control
-        snap.hold_refresh()
-        self.unit.set_workload_version(self.parca_agent.version)
+            logger.exception("Failed to refresh parca-agent snap %s", str(e))
 
-    def _start(self, _):
+    def _on_start(self, _):
         """Start Parca Agent."""
         self.parca_agent.start()
         self.unit.open_port("tcp", 7071)
-        self.unit.status = ops.ActiveStatus()
 
-    def _configure_store(self, event):
-        """Configure remote store with credentials passed over parca-store-endpoint relation."""
-        self.unit.status = ops.MaintenanceStatus("reconfiguring parca-agent")
-        store_config = {} if isinstance(event, RemoveStoreEvent) else event.store_config
-        self.parca_agent.configure(store_config)
-        self.unit.status = ops.ActiveStatus()
-
-    def _remove(self, _):
+    def _on_remove(self, _):
         """Remove Parca Agent from the machine."""
         self.unit.status = ops.MaintenanceStatus("removing parca-agent")
         self.parca_agent.remove()
 
+    def _on_collect_unit_status(self, event: ops.CollectStatusEvent):
+        """Set unit status depending on the state."""
+        if not self.parca_agent.installed:
+            event.add_status(
+                ops.BlockedStatus(
+                    "Failed to install parca-agent snap. Check `juju debug-log` for errors."
+                )
+            )
+
+        # set to blocked if the snap failed to start.
+        # it might happen that the snap would take some time before it becomes "inactive".
+        # if this happens, the charm will be set to blocked in the next processed event.
```

Comment:
maybe let's #FIXME this and add an issue to set up a pebble-checks-like-cronjob to wake up the charm and update its status on parca service changes

Reply:
or look into it at least, not sure how hard that's going to be